cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,238 @@
+      = new SimpleConsumer(brokerInfo.getHost(), brokerInfo.getPort(),
+                           SO_TIMEOUT_MILLIS, BUFFER_SIZE, "simple-kafka-client");
+
+    String clientId = topic + "_" + partition;
+    // Check whether the message fetched with the offset in the given checkpoint has the timestamp from
+    // checkpoint.getNextOffset() - 1 to get the offset corresponding to the timestamp in checkpoint
+    try {
+      long timeStamp = getEventTimeByOffset(consumer, partition, clientId, checkpoint.getNextOffset() - 1);
+      if (timeStamp == checkpoint.getNextEventTime()) {
+        LOG.debug("Checkpoint {} contains matching offset and timestamp. No need to search.", checkpoint);
+        return checkpoint.getNextOffset();
+      }
+    } catch (OffsetOutOfRangeException e) {
+      // Ignore OffsetOutOfRangeException
+    }
+
+    return findNextOffsetByTime(consumer, partition, clientId, checkpoint.getNextEventTime());
+  }
+
+  /**
+   * Determine the lower-bound and upper-bound of offsets to search for the next offset of the message
+   * with smallest offset and log event time equal to {@code targetTime}, and then perform the search.
+   *
+   * @return the matching offset or {@code -1} if not found
+   */
+  private long findNextOffsetByTime(SimpleConsumer consumer, int partition, String clientName, long targetTime)
+    throws IOException {
+    // the offset of the earliest message
+    long minOffset;
+    try {
+      minOffset = KafkaUtil.getOffsetByTimestamp(consumer, topic, partition, targetTime);
+    } catch (IOException e) {
+      LOG.debug("Failed to find the segment with latest message published later than target time {}. " +
+                  "Find the offset with kafka.api.OffsetRequest.EarliestTime() instead.", targetTime, e);
+      minOffset = KafkaUtil.getOffsetByTimestamp(consumer, topic, partition, kafka.api.OffsetRequest.EarliestTime());
+    }
+    // the next offset after the latest message
+    long maxOffset = KafkaUtil.getOffsetByTimestamp(consumer, topic, partition, kafka.api.OffsetRequest.LatestTime());
+    return linearSearchForNextOffset(consumer, partition, clientName, minOffset, maxOffset, targetTime);
+  }
+
+  /**
+   * Performs a linear search to find the next offset of the message with smallest offset and log event time
+   * equal to {@code targetTime}. Stop searching when the current message has log event time
+   * later than {@code maxTime} or offset larger than {@code maxOffset}
+   *
+   * @return next offset of the message with smallest offset and log event time equal to targetTime,
+   *         or next offset of the message with largest offset and timestamp smaller than
+   *         (targetTime - EVENT_DELAY_MILLIS) if no message has log event time equal to targetTime,
+   *         or startOffset if no event has log event time smaller than (targetTime - EVENT_DELAY_MILLIS)
+   */
+  private long linearSearchForNextOffset(SimpleConsumer consumer, int partition, String clientName,
```

Comment:
can you also add in javadoc, why linear search is still reasonable for this scenario.